### PR TITLE
feat: remember last slug for event-type

### LINKED ISSUE #13497: [CAL-3069] remember last event-type slug when renaming
when renaming a slug, it would be nice if the old link still works and redirects

only the previous link, we forget the next one after

i.e.:

cal.com/peer/30min => cal.com/peer/meeting => cal.com/peer/demo = free up /30min, redirect /meeting to /demo

when making a new event, overwrite the redirect

example:

/meeting redirects to /demo

user makes a new /meeting event-type

redirect stops and new /meeting event-type is being used 

<sub>[CAL-3069](https://linear.app/calcom/issue/CAL-3069/remember-last-event-type-slug-when-renaming)</sub>

### PR DESCRIPTION
**Pull request by abhijeetsingh-22**
## What does this PR do?

When changing the `slug` for `event-type` store the previous slug. The previous slug redirects to the current event page.

Fixes #13497 
Fixes CAL-3069
https://github.com/calcom/cal.com/assets/55000107/b7f717ff-18ba-4816-9a7f-c9587d596195

-->

## Type of change

- New feature (non-breaking change which adds functionality)

## How should this be tested?

- Create a event-type 
- Update the slug for the event type.
- Try to access /[user]/[type] with both current and the previous slug. 
- Repeat this for team events.

## Expected Behaviour
Lets say we have a event with slug1
- On changing the slug to slug2 the slug1 should redirect to slug2.
- If a new event type is created with slug1 or any existing event-type is updated to have slug1 then the slug1 stops redirecting to slug2 .

### For managed events
- When users are added to managed event if they have any event-type for which previousSlug is equal to slug of the managed event, then that prevSlug is removed from their event-type.

## Mandatory Tasks

- [ x ] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

## Checklist

- I haven't added tests that prove my fix is effective or that my feature works
- I haven't checked if new and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by PeerRich — COMMENTED** [review]
very cool, will see what priority we can put this in for review

**Comment by zomars** [comment]
My two cents in here. Can we provide an option similarly to this? 


https://github.com/calcom/cal.com/assets/3504472/897fe4a8-9258-476f-bcba-44b5f726231a

**Comment by abhijeetsingh-22** [comment]
https://github.com/calcom/cal.com/assets/55000107/ce87cebc-36ab-4a9c-9d16-bfd4b3fbd8ed

@zomars , pl review I have made the requested changes.

File: apps/web/lib/team/[slug]/[type]/getServerSideProps.tsx [hunk 1]
```diff
@@ -75,6 +75,22 @@ export const getServerSideProps = async (context: GetServerSidePropsContext) =>
...
+      },
+    });
+    if (eventTypeByPreviousSlug) {
+      return {
+        redirect: {
+          destination: `/team/${teamSlug}/${eventTypeByPreviousSlug.slug}`,
+          permanent: true,
+        },
+      };
+    }
```
Comment (abhijeetsingh-22) [inline apps/web/lib/team/[slug]/[type]/getServerSideProps.tsx:93]: If the `event-type` is not found using the slug, try to use `previousSlug` to find the `event-type` if found we should redirect to the slug of the `eventTypeByPreviousSlug`

File: apps/web/modules/users/views/users-type-public-view.getServerSideProps.tsx [hunk 2]
```diff
@@ -160,6 +161,22 @@ async function getUserPageProps(context: GetServerSidePropsContext) {
...
+      },
+    });
+    if (eventTypeByPreviousSlug) {
+      return {
+        redirect: {
+          destination: `/${username}/${eventTypeByPreviousSlug.slug}`,
+          permanent: true,
+        },
+      };
+    }
```
Comment (abhijeetsingh-22) [inline apps/web/modules/users/views/users-type-public-view.getServerSideProps.tsx:179]: Try to find the user `event-type` using `slug`, if not found use the `previousSlug` to find the `event-type` . If the `event-type` is found using the `previousSlug` it means we need to redirect to the `slug` of `eventTypeByPreviousSlug`

File: packages/features/ee/managed-event-types/lib/handleChildrenEventTypes.ts [hunk 3]
```diff
@@ -159,6 +164,10 @@ export default async function handleChildrenEventTypes({
       ? { delete: true }
       : undefined;
   };
+  await removePreviousSlug({
+    userId: [...(newUserIds ?? []), ...(oldUserIds ?? [])],
+    previousSlug: managedEventTypeValues.slug,
+  });
```
Comment (abhijeetsingh-22) [inline packages/features/ee/managed-event-types/lib/handleChildrenEventTypes.ts:170]: Remove the previousSlugs for the `event-type` of users of current event for which `previousSlug === slug`

File: packages/features/eventtypes/lib/previousSlugManager.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+      await prisma.eventType.update({
+        where: updateWhere,
+        data: { previousSlug: null },
+      });
+    } catch (e) {
+      if (!(e instanceof Prisma.PrismaClientKnownRequestError && e.code === "P2025")) {
+        throw new TRPCError({ code: "BAD_REQUEST" });
+      }
+    }
+  }
```
Comment (abhijeetsingh-22) [inline packages/features/eventtypes/lib/previousSlugManager.ts:53]: Used `.update()` method when there is only one userId/teamId as `.update()` is faster than `.updateMany()` for a single entry.

File: packages/lib/server/queries/teams/index.ts [hunk 5]
```diff
@@ -294,7 +294,12 @@ export async function updateNewTeamMemberEventTypes(userId: number, teamId: numb
   });
 
   const allManagedEventTypePropsZod = _EventTypeModel.pick(allManagedEventTypeProps);
-
+  if (eventTypesToAdd.length > 0) {
+    await prisma.eventType.updateMany({
+      where: { userId, previousSlug: { in: eventTypesToAdd.map((e) => e.slug) } },
+      data: { previousSlug: null },
+    });
+  }
```
Comment (abhijeetsingh-22) [inline packages/lib/server/queries/teams/index.ts:302]: When a new user joins the team clear the previousSlugs which are the same as the slug of event-types to add to user.

File: packages/trpc/server/routers/viewer/eventTypes/create.handler.ts [hunk 6]
```diff
@@ -84,6 +86,7 @@ export const createHandler = async ({ ctx, input }: CreateOptions) => {
     };
     data.schedulingType = schedulingType;
   }
+  await removePreviousSlug({ userId, teamId, previousSlug: slug });
```
Comment (abhijeetsingh-22) [inline packages/trpc/server/routers/viewer/eventTypes/create.handler.ts:89]: when creating a new `event-type` remove any `previousSlug` that equals the `slug` in any of the `event-type` of the given user/team.

**Review by keithwillcode — CHANGES_REQUESTED** [review]
Thanks for the contribution! 

Critical features like this around loading booking pages need automated tests written before merge to verify functionality.

**Comment by CarinaWolli** [comment]
Hey @abhijeetsingh-22! Will you be able to fix the merge conflicts and write some tests for it?

**Comment by keithwillcode** [comment]
Closing due to staleness. Please feel free to reopen if work continues.

**Pull request closed without merging**